modules/opengl/include/modules/opengl/shader/shaderobject.h

```diff
@@ -79,41 +96,89 @@ class IVW_MODULE_OPENGL_API ShaderObject {
     void clearShaderExtensions();
 
     /**
-     * \brief adds an additional output specifier to the fragment shader
+     * \brief adds an additional output specifier to the shader
      * The given name will be added as
-     * 
-     *     out vec4 __name__;
-     * 
-     * If location index is positive, the output will be
-     * 
-     *     layout(location = __location__) out vec4 __name__;
-     * 
-     * Location indices can be reused several times unless more than
-     * one output specifier is used.
+     *
+     *     out __type__ __name__;
+     *
+     * The shader will call glBindFragDataLocation for each of the output declarations
+     * before linking the shader
      *
      * @param name      identifier of the output specifier
      * @param location  index location of the output (< MAX_RENDER_TARGETS)
+     * @param type      type used for the output specifier
      */
-    void addOutDeclaration(std::string name, int location = -1);
+    void addOutDeclaration(std::string name, int location = -1, const std::string& type = "vec4");
+    void addOutDeclaration(const OutDeclaration& decl);
     void clearOutDeclarations();
+    const std::vector<OutDeclaration>& getOutDeclarations() const;
+
+    /**
+     * \brief adds an additional input specifier to the shader
+     * The given name will be added as
+     *
+     *     in __type__ __name__;
+     *
+     * The shader will call glBindAttribLocation for each of the input declarations
+     * before linking the shader
+     *
+     * @param name      identifier of the output specifier
+     * @param location  index location of the output (< MAX_RENDER_TARGETS)
+     * @param type      type used for the output specifier
+     */
+    void addInDeclaration(std::string name, int location = -1, const std::string& type = "vec4");
+    void addInDeclaration(const InDeclaration& decl);
+    void clearInDeclarations();
+    const std::vector<InDeclaration>& getInDeclarations() const;
+
+    /**
+     * Adds the default fragment out declarations to the list of out declarations.
+     * This function is automatically called in the constructor for a fragment shader.
+     * If clearOutDeclarations is called then they will be removed and one would have to
+     * manually call this function to re-add them if needed.
+     */
+    void addStandardFragmentOutDeclarations();
```

Comment:
Mention what standard means in our case, i.e.     
```
addOutDeclaration("FragData0", 0);
addOutDeclaration("PickingData", 1);
```
Ideally, this would also be mentioned in the constructor/class documentation